[WIP] Buffs rework

superewald opened the pull request:
> **This is a WIP and still needs a few bugfixes and improvements but I'm looking forward to suggestions already :)**

The buff system is currently lacking a bit functionality and a lot of code style compliance which this PR solves.

## changelist

1. introduced `IBuffManager` and `BuffManager` to decouple buff logic from `ObjAIBase` (#1001)
   > *all buff related functions are now accessable in `IObjAIBase` through `Buffs`*
   > *for eg:`aiBase.Buffs.Add()`*
3. introduced flags for buffs to mark a buff as not interruptable by cc for eg. (#973)
4. introduced a real priority queue for Buffs (#886)

## small improvements
| what                                               | why                                  | where        |
| -------------------------------------------------- | ------------------------------------ | ------------ |

## temporary buglist

*These are curren't known bugs (as of WIP).*

1. last stack of overlapping buff not removed
1. recall animation is 4 seconds long instead of 8
1. still can cast spell while recall/tping

*Refs:* #1001, #973, #886

Commit:
[WIP-COMMIT]
- added BuffManager and IBuffManager
- removed old buff logic form ObjAIBase
- added BuffFlags
- changed functions accessing buff logic

Commit:
- removed AddStatModifier/RemoveStatModifier from ObjAIBase & BuffManager since already in ObjAIBase.Stats
- fixed buff/spell scripts

Commit:
- added comments to IBuffManager
- added initialization of values in BuffManager

Commit:
- added missing variable initialization
- added comments

Commit:
**temporary commit**: should resolve #886 but crashes atm
- added priority queue for buffs (based on order & duration)

Commit:
added missing code

Commit:
- fix GetSlot code to properly generate new slot (which lead to client crash before)

Commit:
- fixed buff dequeuing

superewald commented on GameServerCore/Domain/GameObjects/IObjAiBase.cs:
```diff
@@ -20,28 +20,16 @@ public interface IObjAiBase : IAttackableUnit
         bool IsMelee { get; set; }
         MoveOrder MoveOrder { get; }
         IAttackableUnit TargetUnit { get; set;  }
+        IBuffManager Buffs { get; }
```
Maybe we should move this to IAttackableUnit so Turrets and other attackable objects can get buffs.

superewald replied:
maybe move this to `IAttackbleUnit` so other objects like turrets can get buffed.

superewald commented on GameServerCore/Domain/GameObjects/IObjAiBase.cs:
```diff
@@ -20,28 +20,16 @@ public interface IObjAiBase : IAttackableUnit
-        void AddBuff(IBuff b);
-        void AddStatModifier(IStatsModifier statModifier);
+        
         void ApplyCrowdControl(ICrowdControl cc);
```
remove that stuff and implement it as plain buff ?

superewald commented on GameServerCore/Domain/IBuff.cs:
```diff
@@ -23,5 +24,8 @@ public interface IBuff : IStackable, IUpdate
         void SetSlot(byte slot);
+
+        bool HasFlag(BuffFlag flag);
+        bool HasFlags(List<BuffFlag> flags);
```
*self-note: use array instead of list*

superewald commented on GameServerLib/API/ApiFunctionManager.cs:
```diff
@@ -99,26 +99,26 @@ public static IBuff AddBuff(string buffName, float duration, byte stacks, ISpell
 
             try
             {
-                buff = new Buff(_game, buffName, duration, stacks, originspell, onto, from, infiniteduration);
```
*self-note: revert this line*

superewald commented on GameServerLib/ObjectManager.cs:
```diff
@@ -134,16 +139,18 @@ public void Update(float diff)
                 {
-                    var tempBuffs = ai.GetBuffs();
-                    for (int i = tempBuffs.Count - 1; i >= 0; i--)
+                    var tempBuffs = ai.Buffs.Get();  
```
I don't know if `SimplePriorityQueue.ToList()` preserves order of elements, but I can't seem to implement a propper dequeue-loop since `BuffManager._buffsQueue` gets modified by changing `tempBuffs` (even when shadow-copy the `_buffsQueue`).

danil179 replied:
It seems from here https://github.com/BlueRaja/High-Speed-Priority-Queue-for-C-Sharp/blob/master/Priority%20Queue/GenericPriorityQueue.cs that the order is like a perfect tree implemented by an array. That means that the order is completely random more or less, not even sorted by priority.

superewald replied:
Well, im stupid. Instead of using a prio queue internaly, i can just create it on demand...

danil179 commented on GameServerLib/GameObjects/Spells/BuffManager.cs:
```diff
@@ -0,0 +1,410 @@
+        //private Dictionary<string, IBuff> _buffs;
+        private SimplePriorityQueue<IBuff, float> _buffQueue;
+        private IBuff[] _slots;
+        private Game _game;
```
If you can find a way to avoid this `Game` class get copied everywhere, please do it. I have a suggestion to just copy `PacketNotifier` this time.

danil179 reviewed (commented):
A few minor comments.

danil179 commented:
Due to the last PRs you need to rebase, and just change the code in ProtectionManager `AddStatModifier->Stats.Add...`.

Commit:
[temp] needs a few refactors
- now uses real dequeueing instead of foreach

Commit:
- renamed variables to match content

Commit:
- removed old comments

Commit:
- removed _game from BuffManager
- updated ProtectionManager to resolve latest prs

Commit:
- removed debug message
- added recall particle lifetime

Commit:
- added more ways to interrupt recall

danil179 commented on GameServerLib/ObjectManager.cs:
```diff
@@ -134,16 +139,17 @@ public void Update(float diff)
                 {
-                    var tempBuffs = ai.GetBuffs();
-                    for (int i = tempBuffs.Count - 1; i >= 0; i--)
+                    var buffQueue = ai.Buffs.GetQueue();  
```
Remove 2 spaces after `;`.

danil179 commented on GameServerCore/Domain/IBuffManager.cs:
```diff
@@ -0,0 +1,156 @@
+        /// get the buff priority queue
+        /// </summary>
+        /// <returns>buff priority queue</returns>
+        SimplePriorityQueue<IBuff, float> GetQueue();
```
I think that it is incorrect to expose implementation here. I don't see any need of using queue. The `BuffManger` can have a function like `GetNextElapsedBuff` by inspecting the buff in the top of the queue and then just update all the buffs (and remove those with time that is before/equal to this elapsed time).

danil179 commented on GameServerLib/ObjectManager.cs:
```diff
@@ -134,16 +139,17 @@ public void Update(float diff)
-                    var tempBuffs = ai.GetBuffs();
-                    for (int i = tempBuffs.Count - 1; i >= 0; i--)
+                    var buffQueue = ai.Buffs.GetQueue();  
+                    while(buffQueue.Count > 0)
```
If you already iterate all the buffs to update them, there is no point in using the queue, as you spend greater computational resources on the building of the queue (O(n)/O(nlog n) depends on implementation) and on the constant dequeue (which is usually O(log(n))). Ultimately, this will take much more time than O(n) checks.

The reason for a queue is to identify buffs that needs to be removed faster in case this will be needed, and to handle the priority in sort of pipeline (so FIFO ordering which also allows us to know the elapsed times in order, as it is crucial for some case uses).

This current loops is very inefficient.

danil179 replied:
Just to add some more info. Many buffs don't even need `udate` and their `OnUpdate` is simply empty. We should at some point (not now) keep track on these types of buffs by probably storing them in a separate queue.

superewald replied:
You mean like this ?

```cs
        public void Update(float diff)
        {
            lock(_buffQueueLock)
            {
                // remove elapsed buffs
                while (_buffQueue.Count != 0 && _buffQueue.GetPriority(_buffQueue.First) <= 0)
                {
                    Remove(_buffQueue.First);
                }

                // update remaining buffs
                foreach (var buff in _buffQueue)
                {
                    _buffQueue.UpdatePriority(buff, _buffQueue.GetPriority(buff) - (diff / 1000));
                    buff.Update(diff);
                }
            }
        }
```

danil179 reviewed (commented):
Some comments.

Commit:
- fixed buffs didnt get removed

Commit:
- added BuffManager.Update
- queue uses priority of remaining buff time & insertion order

Commit:
- fixed api remove buff
- fixed crash on stacked buffs
- (bug) last stack of stackable buff not removed

Commit:
- fixed stackable buffs
- removed debug messages

superewald commented:
I'm gonna rewrite this PR since this one contains too much unnecessary commits and changes.

superewald commented:
Closed due to dirty implementations and too much commits. 
*Continued by*:  #1006
